cmd2/cmd2.py

```diff
@@ -2006,24 +2028,65 @@ def onecmd(self, statement: Union[Statement, str]) -> Optional[bool]:
         if not isinstance(statement, Statement):
             statement = self._complete_statement(statement)
 
-        funcname = self._func_named(statement.command)
-        if not funcname:
-            self.default(statement)
-            return
+        # Check if this is a macro
+        if statement.command in self.macros:
+            stop = self._run_macro(statement)
+        else:
+            funcname = self._func_named(statement.command)
+            if not funcname:
+                self.default(statement)
+                return False
 
-        # Since we have a valid command store it in the history
-        if statement.command not in self.exclude_from_history:
-            self.history.append(statement.raw)
+            # Since we have a valid command store it in the history
+            if statement.command not in self.exclude_from_history:
+                self.history.append(statement.raw)
 
-        try:
-            func = getattr(self, funcname)
-        except AttributeError:
-            self.default(statement)
-            return
+            try:
+                func = getattr(self, funcname)
+            except AttributeError:
+                self.default(statement)
+                return False
+
+            stop = func(statement)
 
-        stop = func(statement)
         return stop
 
+    def _run_macro(self, statement: Statement) -> bool:
+        """
+        Resolves a macro and runs the resulting string
+
+        :param statement: the parsed statement from the command line
+        :return: a flag indicating whether the interpretation of commands should stop
+        """
+        try:
+            macro = self.macros[statement.command]
+        except KeyError:
+            raise KeyError("{} is not a macro".format(statement.command))
+
+        # Confirm we have the correct number of arguments
+        if len(statement.arg_list) != macro.required_arg_count:
```

Comment:
Currently I require the exact number of expected arguments, since its like calling a function.
Should I allow extra arguments and just append them to the end of the resolved command?

Reply:
No, I think your approach is good. Just add a comment.

```Python
    # For macros, every argument must be provided and there can be no extra arguments.
```